cmd/kube-apiserver/apiserver.go

```diff
@@ -20,6 +20,7 @@ package main
 
 import (
 	"os"
+	_ "time/tzdata" // for timeZone support in CronJob
```

Comment:
This is only <500kb, but I want to note that we're likely not using this copy anyhow, since the images we publish will have tzdata in them (and .. pretty much any docker base image short of literal `FROM SCRATCH`) and if the binary is run directly on the host (which seems unusual) host environments generally will as well.

Note from package docs:
>. Package tzdata provides an embedded copy of the timezone database. If this package is imported anywhere in the program, then if the time package cannot find tzdata files on the system, it will use this embedded information


xref: https://github.com/kubernetes/enhancements/pull/3143#discussion_r837766209

Reply:
It's true that it will probably import the tzdata from the image, but I think it's also important to spell out dependencies when possible. Also, I've had trouble in the past with Alpine and other minimal images not including that data and ultimately breaking things.

Reply:
> It's true that it will probably import the tzdata from the image, but I think it's also important to spell out dependencies when possible.

It's not clear to me that embedded tzdata is a real dependency that we desire, it will increase the size of our images (granted not drastically) for zero gain since all of our base images contain tzdata, even "distroless", AFAICT there is no scenario in the entire Kubernetes organization + subprojects where this data will actually be used.

> Also, I've had trouble in the past with Alpine and other minimal images not including that data and ultimately breaking things.

Yes, but we don't ship alpine based images in Kubernetes, and anyone doing their own custom images is liable to break via failing to include any number of packages already (e.g. cacerts) and should probably just install tzdata.

Reply:
It also means we have another place where Kubernetes should be trying to keep tzdata patched in the releases we ship, I'm not sure this is desirable vs just using tzdata from the docker images, not to mention being decoupled from go version lifecycle ... 🤔

Reply:
ACK, though removing it would be itself would potentially be a breaking change for users.

I feel like the expectations around tzdata freshness should have been outlined in the KEP, commented there. I realize this is a bit late.

Reply:
lack of available/uniform tzdata was a blocker for this feature for a long time. I'm strongly in favor of including this fallback to ensure we always have data available.

Reply:
(if we want to put it in a build-flag gated file so someone can exclude it if they know they don't need it, I wouldn't object to that)

Reply:
Correct what Jordan wrote above, last time we discussed and rejected further implementation of this feature was because we did not want to force users to install tzdata explicitly. Having this mechanism as a fallback is what allowed us to proceed. This was explicitly laid out in the motivation section of the KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3140-TimeZone-support-in-CronJob#motivation

Reply:
Further more, since this tzdata is embedded inside golang, every time we patch golang we get appropriate patches with it, so I'm not worried about it, since it does not require separate release cadence. Also, as you mentioned the 500kb barely makes a difference in the size of 100MB+ binaries such as kube-apisever and kube-controller-manager.

Reply:
Golang patch releases do not necessarily happen when the tzdb updates, and go releases are not supported with new patches after some time so users depending on this will be on their own.

I still think it would be reasonable to require tzdb to run the apiserver binaries we ship anyhow, we don't even test it without running in our images which contain it and we require other reasonable system packages to function.

If we make the expectations clear that this will not receive special patching vs regular golang upgrade cadence, I think it is reasonable though.

Reply:
I don't think this deserves a build flag for exclusion fwiw. I just don't want any expectation that outdated bundled tzdata will lead to urgent golang updates + k8s releases and I don't see this spelled out anywhere. If we didn't provide the fallback the stance is clearer that users are responsible for having an updated tzdb in their local install. I also don't think we exercise this fallback in CI FWIW though we can probably trust golang there.